File: chromium/chrome/browser/safety_check/android/safety_check_bridge.cc
// Copyright 2020 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "chrome/browser/safety_check/android/safety_check_bridge.h"

#include <jni.h>

#include "chrome/browser/profiles/profile_manager.h"
#include "chrome/browser/safety_check/android/jni_headers/SafetyCheckBridge_jni.h"
#include "components/safety_check/safety_check.h"

static jlong JNI_SafetyCheckBridge_Init(
    JNIEnv* env,
    const base::android::JavaParamRef<jobject>& obj,
    const base::android::JavaParamRef<jobject>& j_safety_check_observer) {
  return reinterpret_cast<intptr_t>(
      new SafetyCheckBridge(env, j_safety_check_observer));
}

SafetyCheckBridge::SafetyCheckBridge(
    JNIEnv* env,
    const base::android::JavaParamRef<jobject>& j_safety_check_observer)
    : pref_service_(ProfileManager::GetActiveUserProfile()
                        ->GetOriginalProfile()
                        ->GetPrefs()),
      j_safety_check_observer_(j_safety_check_observer) {
  safety_check_.reset(new safety_check::SafetyCheck(this));
  password_check_controller_.reset(new BulkLeakCheckControllerAndroid());
}

void SafetyCheckBridge::Destroy(
    JNIEnv* env,
    const base::android::JavaParamRef<jobject>& obj) {
  password_check_controller_.reset();
  safety_check_.reset();
  delete this;
}

void SafetyCheckBridge::CheckSafeBrowsing(
    JNIEnv* env,
    const base::android::JavaParamRef<jobject>& obj) {
  safety_check_->CheckSafeBrowsing(pref_service_);
}

void SafetyCheckBridge::CheckPasswords(
    JNIEnv* env,
    const base::android::JavaParamRef<jobject>& obj) {
  password_check_controller_->AddObserver(this);
  password_check_controller_->StartPasswordCheck();
}

int SafetyCheckBridge::GetNumberOfPasswordLeaksFromLastCheck(
    JNIEnv* env,
    const base::android::JavaParamRef<jobject>& obj) {
  return password_check_controller_->GetNumberOfLeaksFromLastCheck();
}

bool SafetyCheckBridge::SavedPasswordsExist(
    JNIEnv* env,
    const base::android::JavaParamRef<jobject>& obj) {
  return password_check_controller_->GetNumberOfSavedPasswords() != 0;
}

void SafetyCheckBridge::StopObservingPasswordsCheck(
    JNIEnv* env,
    const base::android::JavaParamRef<jobject>& obj) {
  password_check_controller_->RemoveObserver(this);
}

void SafetyCheckBridge::OnSafeBrowsingCheckResult(
    safety_check::SafetyCheck::SafeBrowsingStatus status) {
  JNIEnv* env = base::android::AttachCurrentThread();
  Java_SafetyCheckCommonObserver_onSafeBrowsingCheckResult(
      env, j_safety_check_observer_, static_cast<int>(status));
}

void SafetyCheckBridge::OnStateChanged(
    password_manager::BulkLeakCheckService::State state) {
  JNIEnv* env = base::android::AttachCurrentThread();
  Java_SafetyCheckCommonObserver_onPasswordCheckStateChange(
      env, j_safety_check_observer_, static_cast<int>(state));
}

void SafetyCheckBridge::OnCredentialDone(
    const password_manager::LeakCheckCredential& credential,
    password_manager::IsLeaked is_leaked,
    SafetyCheckBridge::DoneCount credentials_checked,
    SafetyCheckBridge::TotalCount total_to_check) {
  JNIEnv* env = base::android::AttachCurrentThread();
  Java_SafetyCheckCommonObserver_onPasswordCheckCredentialDone(
      env, j_safety_check_observer_, credentials_checked.value(),
      total_to_check.value());
}

SafetyCheckBridge::~SafetyCheckBridge() = default;

File: chromium/chrome/browser/safety_check/android/java/src/org/chromium/chrome/browser/safety_check/SafetyCheckBridge.java
// Copyright 2020 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.chrome.browser.safety_check;

import org.chromium.base.annotations.CalledByNative;
import org.chromium.base.annotations.NativeMethods;
import org.chromium.chrome.browser.password_check.BulkLeakCheckServiceState;

/**
 * Provides access to the C++ multi-platform Safety check code in
 * //components/safety_check.
 */
public class SafetyCheckBridge {
    /**
     * Observer for SafetyCheck code common to Desktop, Android, and iOS.
     */
    public interface SafetyCheckCommonObserver {
        /**
         * Gets invoked by the C++ code once the Safe Browsing check has results.
         *
         * @param status SafetyCheck::SafeBrowsingStatus enum value representing the
         *               Safe Browsing state (see
         *               //components/safety_check/safety_check.h).
         */
        @CalledByNative("SafetyCheckCommonObserver")
        void onSafeBrowsingCheckResult(@SafeBrowsingStatus int status);

        /**
         * Gets invoked by the C++ code every time another credential is checked.
         *
         * @param checked Number of passwords already checked.
         * @param total   Total number of passwords to check.
         */
        @CalledByNative("SafetyCheckCommonObserver")
        void onPasswordCheckCredentialDone(int checked, int total);

        /**
         * Gets invoked by the C++ code when the status of the password check changes.
         *
         * @param state BulkLeakCheckService::State enum value representing the state
         *              (see //components/password_manager/core/browser
         *                  /bulk_leak_check_service_interface.h).
         */
        @CalledByNative("SafetyCheckCommonObserver")
        void onPasswordCheckStateChange(@BulkLeakCheckServiceState int state);
    }

    /**
     * Holds the C++ side of the Bridge class.
     */
    private long mNativeSafetyCheckBridge;

    /**
     * Initializes the C++ side.
     *
     * @param safetyCheckCommonObserver An observer instance that will receive the
     *                                  result of the check.
     */
    public SafetyCheckBridge(SafetyCheckCommonObserver safetyCheckCommonObserver) {
        mNativeSafetyCheckBridge =
                SafetyCheckBridgeJni.get().init(SafetyCheckBridge.this, safetyCheckCommonObserver);
    }

    /**
     * Triggers the Safe Browsing check on the C++ side.
     */
    void checkSafeBrowsing() {
        SafetyCheckBridgeJni.get().checkSafeBrowsing(
                mNativeSafetyCheckBridge, SafetyCheckBridge.this);
    }

    /**
     * Triggers the passwords check on the C++ side.
     */
    void checkPasswords() {
        SafetyCheckBridgeJni.get().checkPasswords(mNativeSafetyCheckBridge, SafetyCheckBridge.this);
    }

    /**
     * @return Number of password leaks discovered during the last check.
     */
    int getNumberOfPasswordLeaksFromLastCheck() {
        return SafetyCheckBridgeJni.get().getNumberOfPasswordLeaksFromLastCheck(
                mNativeSafetyCheckBridge, SafetyCheckBridge.this);
    }

    /**
     * @return Whether the user has some passwords saved.
     */
    boolean savedPasswordsExist() {
        return SafetyCheckBridgeJni.get().savedPasswordsExist(
                mNativeSafetyCheckBridge, SafetyCheckBridge.this);
    }

    /**
     * Stops observing the events of the passwords leak check.
     */
    void stopObservingPasswordsCheck() {
        SafetyCheckBridgeJni.get().stopObservingPasswordsCheck(
                mNativeSafetyCheckBridge, SafetyCheckBridge.this);
    }

    /**
     * Destroys the C++ side of the Bridge, freeing up all the associated memory.
     */
    void destroy() {
        assert mNativeSafetyCheckBridge != 0;
        SafetyCheckBridgeJni.get().destroy(mNativeSafetyCheckBridge, SafetyCheckBridge.this);
        mNativeSafetyCheckBridge = 0;
    }

    /**
     * C++ method signatures.
     */
    @NativeMethods
    interface Natives {
        long init(SafetyCheckBridge safetyCheckBridge, SafetyCheckCommonObserver observer);
        void checkSafeBrowsing(long nativeSafetyCheckBridge, SafetyCheckBridge safetyCheckBridge);
        void checkPasswords(long nativeSafetyCheckBridge, SafetyCheckBridge safetyCheckBridge);
        int getNumberOfPasswordLeaksFromLastCheck(
                long nativeSafetyCheckBridge, SafetyCheckBridge safetyCheckBridge);
        boolean savedPasswordsExist(
                long nativeSafetyCheckBridge, SafetyCheckBridge safetyCheckBridge);
        void stopObservingPasswordsCheck(
                long nativeSafetyCheckBridge, SafetyCheckBridge safetyCheckBridge);
        void destroy(long nativeSafetyCheckBridge, SafetyCheckBridge safetyCheckBridge);
    }
}
